<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=9"/>
<meta name="generator" content="Doxygen 1.8.6"/>
<title>mapgd: models Class Reference</title>
<link href="tabs.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="jquery.js"></script>
<script type="text/javascript" src="dynsections.js"></script>
<link href="search/search.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="search/search.js"></script>
<script type="text/javascript">
  $(document).ready(function() { searchBox.OnSelectItem(0); });
</script>
<link href="doxygen.css" rel="stylesheet" type="text/css" />
</head>
<body>
<div id="top"><!-- do not remove this div, it is closed by doxygen! -->
<div id="titlearea">
<table cellspacing="0" cellpadding="0">
 <tbody>
 <tr style="height: 56px;">
  <td style="padding-left: 0.5em;">
   <div id="projectname">mapgd
   &#160;<span id="projectnumber">0.4</span>
   </div>
   <div id="projectbrief">A program for the Maximum-likelihood analysis of population genomic data.</div>
  </td>
 </tr>
 </tbody>
</table>
</div>
<!-- end header part -->
<!-- Generated by Doxygen 1.8.6 -->
<script type="text/javascript">
var searchBox = new SearchBox("searchBox", "search",false,'Search');
</script>
  <div id="navrow1" class="tabs">
    <ul class="tablist">
      <li><a href="index.html"><span>Main&#160;Page</span></a></li>
      <li><a href="pages.html"><span>Related&#160;Pages</span></a></li>
      <li><a href="modules.html"><span>Modules</span></a></li>
      <li class="current"><a href="annotated.html"><span>Data&#160;Structures</span></a></li>
      <li><a href="files.html"><span>Files</span></a></li>
      <li>
        <div id="MSearchBox" class="MSearchBoxInactive">
        <span class="left">
          <img id="MSearchSelect" src="search/mag_sel.png"
               onmouseover="return searchBox.OnSearchSelectShow()"
               onmouseout="return searchBox.OnSearchSelectHide()"
               alt=""/>
          <input type="text" id="MSearchField" value="Search" accesskey="S"
               onfocus="searchBox.OnSearchFieldFocus(true)" 
               onblur="searchBox.OnSearchFieldFocus(false)" 
               onkeyup="searchBox.OnSearchFieldChange(event)"/>
          </span><span class="right">
            <a id="MSearchClose" href="javascript:searchBox.CloseResultsWindow()"><img id="MSearchCloseImg" border="0" src="search/close.png" alt=""/></a>
          </span>
        </div>
      </li>
    </ul>
  </div>
  <div id="navrow2" class="tabs2">
    <ul class="tablist">
      <li><a href="annotated.html"><span>Data&#160;Structures</span></a></li>
      <li><a href="classes.html"><span>Data&#160;Structure&#160;Index</span></a></li>
      <li><a href="inherits.html"><span>Class&#160;Hierarchy</span></a></li>
      <li><a href="functions.html"><span>Data&#160;Fields</span></a></li>
    </ul>
  </div>
<!-- window showing the filter options -->
<div id="MSearchSelectWindow"
     onmouseover="return searchBox.OnSearchSelectShow()"
     onmouseout="return searchBox.OnSearchSelectHide()"
     onkeydown="return searchBox.OnSearchSelectKey(event)">
<a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(0)"><span class="SelectionMark">&#160;</span>All</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(1)"><span class="SelectionMark">&#160;</span>Data Structures</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(2)"><span class="SelectionMark">&#160;</span>Functions</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(3)"><span class="SelectionMark">&#160;</span>Variables</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(4)"><span class="SelectionMark">&#160;</span>Friends</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(5)"><span class="SelectionMark">&#160;</span>Groups</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(6)"><span class="SelectionMark">&#160;</span>Pages</a></div>

<!-- iframe showing the search results (closed by default) -->
<div id="MSearchResultsWindow">
<iframe src="javascript:void(0)" frameborder="0" 
        name="MSearchResults" id="MSearchResults">
</iframe>
</div>

</div><!-- top -->
<div class="header">
  <div class="summary">
<a href="#pub-methods">Public Member Functions</a> &#124;
<a href="#pri-attribs">Private Attributes</a>  </div>
  <div class="headertitle">
<div class="title">models Class Reference</div>  </div>
</div><!--header-->
<div class="contents">

<p><code>#include &lt;<a class="el" href="models_8h_source.html">models.h</a>&gt;</code></p>
<div class="dynheader">
Collaboration diagram for models:</div>
<div class="dyncontent">
<div class="center"><img src="classmodels__coll__graph.png" border="0" usemap="#models_coll__map" alt="Collaboration graph"/></div>
<map name="models_coll__map" id="models_coll__map">
<area shape="rect" id="node2" href="classlnmultinomial.html" title="lnmultinomial" alt="" coords="5,5,104,32"/></map>
<center><span class="legend">[<a href="graph_legend.html">legend</a>]</span></center></div>
<table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a name="pub-methods"></a>
Public Member Functions</h2></td></tr>
<tr class="memitem:a7bbad4e9015444af15acbbd736f690ed"><td class="memItemLeft" align="right" valign="top">float_t&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classmodels.html#a7bbad4e9015444af15acbbd736f690ed">loglikelihood</a> (const <a class="el" href="classLocus.html">Locus</a> &amp;, const <a class="el" href="classAllele.html">Allele</a> &amp;)</td></tr>
<tr class="memdesc:a7bbad4e9015444af15acbbd736f690ed"><td class="mdescLeft">&#160;</td><td class="mdescRight">Gets the log likelihood of the observations at <a class="el" href="classLocus.html">Locus</a>, given <a class="el" href="classAllele.html" title="Summary statistics from the allele command. ">Allele</a>.  <a href="#a7bbad4e9015444af15acbbd736f690ed">More...</a><br/></td></tr>
<tr class="separator:a7bbad4e9015444af15acbbd736f690ed"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a48a314cea7f2ada0dd705ff5fb2814d6"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a48a314cea7f2ada0dd705ff5fb2814d6"></a>
float_t&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classmodels.html#a48a314cea7f2ada0dd705ff5fb2814d6">goodness_of_fit</a> (<a class="el" href="classLocus.html">Locus</a> &amp;, const <a class="el" href="classAllele.html">Allele</a> &amp;, std::vector&lt; float_t &gt; &amp;, const float_t &amp;)</td></tr>
<tr class="memdesc:a48a314cea7f2ada0dd705ff5fb2814d6"><td class="mdescLeft">&#160;</td><td class="mdescRight">Gets the log likelihood of the observations at <a class="el" href="classLocus.html">Locus</a>, given <a class="el" href="classAllele.html" title="Summary statistics from the allele command. ">Allele</a>. <br/></td></tr>
<tr class="separator:a48a314cea7f2ada0dd705ff5fb2814d6"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:af826c332f07d74c4750302f934ccaa09"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="af826c332f07d74c4750302f934ccaa09"></a>
void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classmodels.html#af826c332f07d74c4750302f934ccaa09">init_gof</a> (const count_t *, const <a class="el" href="classAllele.html">Allele</a> &amp;)</td></tr>
<tr class="memdesc:af826c332f07d74c4750302f934ccaa09"><td class="mdescLeft">&#160;</td><td class="mdescRight">breif Initilizes the ? for goodness of fit calculations. <br/></td></tr>
<tr class="separator:af826c332f07d74c4750302f934ccaa09"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a6812b41b60a089d7d113cf863b115647"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a6812b41b60a089d7d113cf863b115647"></a>
float_t&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classmodels.html#a6812b41b60a089d7d113cf863b115647">get_gof</a> (const count_t *, const <a class="el" href="classAllele.html">Allele</a> &amp;)</td></tr>
<tr class="memdesc:a6812b41b60a089d7d113cf863b115647"><td class="mdescLeft">&#160;</td><td class="mdescRight">Returns the likelihood of a ?? goodness of fit... blah blah blah. <br/></td></tr>
<tr class="separator:a6812b41b60a089d7d113cf863b115647"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a379f9643a0899bfe81af18842c25356b"><td class="memItemLeft" align="right" valign="top">float_t&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classmodels.html#a379f9643a0899bfe81af18842c25356b">genotypelikelihood</a> (const <a class="el" href="structquartet.html">quartet_t</a> &amp;, const <a class="el" href="classAllele.html">Allele</a> &amp;)</td></tr>
<tr class="memdesc:a379f9643a0899bfe81af18842c25356b"><td class="mdescLeft">&#160;</td><td class="mdescRight">Clalculates genotypic likelihoods. Not implement, may be depricated.  <a href="#a379f9643a0899bfe81af18842c25356b">More...</a><br/></td></tr>
<tr class="separator:a379f9643a0899bfe81af18842c25356b"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a4c3e1316da574d7eab691e8dfd32c486"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a4c3e1316da574d7eab691e8dfd32c486"></a>
float_t&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classmodels.html#a4c3e1316da574d7eab691e8dfd32c486">lnL</a> (const float_t &amp;logMM, const float_t &amp;logMm, const float &amp;logmm, const count_t *count)</td></tr>
<tr class="memdesc:a4c3e1316da574d7eab691e8dfd32c486"><td class="mdescLeft">&#160;</td><td class="mdescRight">Clalculates goodness of fit likelihoods. <br/></td></tr>
<tr class="separator:a4c3e1316da574d7eab691e8dfd32c486"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a33b232eed7d936124a470a2c87ffdc39"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a33b232eed7d936124a470a2c87ffdc39"></a>
<a class="el" href="classmodels.html">models</a> &amp;&#160;</td><td class="memItemRight" valign="bottom"><b>operator=</b> (const <a class="el" href="classmodels.html">models</a> &amp;rhs)</td></tr>
<tr class="separator:a33b232eed7d936124a470a2c87ffdc39"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table><table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a name="pri-attribs"></a>
Private Attributes</h2></td></tr>
<tr class="memitem:a8366e06c832207001f9fa351315e92e5"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a8366e06c832207001f9fa351315e92e5"></a>
<a class="el" href="classlnmultinomial.html">lnmultinomial</a>&#160;</td><td class="memItemRight" valign="bottom"><b>lnMM_</b></td></tr>
<tr class="separator:a8366e06c832207001f9fa351315e92e5"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a1961bd182932816bf4c898f4aab78f35"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a1961bd182932816bf4c898f4aab78f35"></a>
<a class="el" href="classlnmultinomial.html">lnmultinomial</a>&#160;</td><td class="memItemRight" valign="bottom"><b>lnMm_</b></td></tr>
<tr class="separator:a1961bd182932816bf4c898f4aab78f35"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a006e767f301f845494ffabb72fdaae2c"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a006e767f301f845494ffabb72fdaae2c"></a>
<a class="el" href="classlnmultinomial.html">lnmultinomial</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classmodels.html#a006e767f301f845494ffabb72fdaae2c">lnmm_</a></td></tr>
<tr class="memdesc:a006e767f301f845494ffabb72fdaae2c"><td class="mdescLeft">&#160;</td><td class="mdescRight">The three multinomials we will use for probability calculations. <br/></td></tr>
<tr class="separator:a006e767f301f845494ffabb72fdaae2c"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a7704b5b116e535792ad793c4196ad7b8"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a7704b5b116e535792ad793c4196ad7b8"></a>
<a class="el" href="classlnmultinomial.html">lnmultinomial</a>&#160;</td><td class="memItemRight" valign="bottom"><b>lnMMP_</b></td></tr>
<tr class="separator:a7704b5b116e535792ad793c4196ad7b8"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a29b19232203e76420ed3bd7a62e6303b"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a29b19232203e76420ed3bd7a62e6303b"></a>
<a class="el" href="classlnmultinomial.html">lnmultinomial</a>&#160;</td><td class="memItemRight" valign="bottom"><b>lnMmP_</b></td></tr>
<tr class="separator:a29b19232203e76420ed3bd7a62e6303b"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:aa41aca7f89f1b31aac86390a70f32b84"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="aa41aca7f89f1b31aac86390a70f32b84"></a>
<a class="el" href="classlnmultinomial.html">lnmultinomial</a>&#160;</td><td class="memItemRight" valign="bottom"><b>lnmmP_</b></td></tr>
<tr class="separator:aa41aca7f89f1b31aac86390a70f32b84"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a4e3600d74dd1d821d3883e7b1b865880"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a4e3600d74dd1d821d3883e7b1b865880"></a>
float_t&#160;</td><td class="memItemRight" valign="bottom"><b>E0_</b></td></tr>
<tr class="separator:a4e3600d74dd1d821d3883e7b1b865880"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ab9cbc7e63358457207a99e8f789e018b"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="ab9cbc7e63358457207a99e8f789e018b"></a>
float_t&#160;</td><td class="memItemRight" valign="bottom"><b>E1_</b></td></tr>
<tr class="separator:ab9cbc7e63358457207a99e8f789e018b"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a1dc72dc854b102c0b8024c2f43633d39"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a1dc72dc854b102c0b8024c2f43633d39"></a>
float_t&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classmodels.html#a1dc72dc854b102c0b8024c2f43633d39">E2_</a></td></tr>
<tr class="memdesc:a1dc72dc854b102c0b8024c2f43633d39"><td class="mdescLeft">&#160;</td><td class="mdescRight">Values used in calculations;. <br/></td></tr>
<tr class="separator:a1dc72dc854b102c0b8024c2f43633d39"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table>
<a name="details" id="details"></a><h2 class="groupheader">Detailed Description</h2>
<div class="textblock"><p>this class perfoms some likelihood calculations. It keeps some look-up tables that will grow as calculations are performed, and may become to large over time. TODO: implement automatic look-up table purging.</p>
<p>Currently likelihood calculations are all performed by a set of 'models' that we give to the "multinomial" class via the multinomial::set method. These models should be able to look at the <a class="el" href="classAllele.html" title="Summary statistics from the allele command. ">Allele</a> structure, which contains information about the error rate at the locus and the identity of the major and minor allele, and return a set of four log probabilities of observing each particular nucleotide in a given call. </p>
</div><h2 class="groupheader">Member Function Documentation</h2>
<a class="anchor" id="a379f9643a0899bfe81af18842c25356b"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">float_t models::genotypelikelihood </td>
          <td>(</td>
          <td class="paramtype">const <a class="el" href="structquartet.html">quartet_t</a> &amp;&#160;</td>
          <td class="paramname"><em>quartet</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const <a class="el" href="classAllele.html">Allele</a> &amp;&#160;</td>
          <td class="paramname"><em>population</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Clalculates genotypic likelihoods. Not implement, may be depricated. </p>
<p>DONT USE THIS!!! TODO FIX IT!. </p>

</div>
</div>
<a class="anchor" id="a7bbad4e9015444af15acbbd736f690ed"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">float_t models::loglikelihood </td>
          <td>(</td>
          <td class="paramtype">const <a class="el" href="classLocus.html">Locus</a> &amp;&#160;</td>
          <td class="paramname"><em>site</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const <a class="el" href="classAllele.html">Allele</a> &amp;&#160;</td>
          <td class="paramname"><em>p</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Gets the log likelihood of the observations at <a class="el" href="classLocus.html">Locus</a>, given <a class="el" href="classAllele.html" title="Summary statistics from the allele command. ">Allele</a>. </p>
<p>A function that calculates the logl likelihood of a set of observations. </p>

</div>
</div>
<hr/>The documentation for this class was generated from the following files:<ul>
<li>mapgd_0.4/math_functions/<a class="el" href="models_8h_source.html">models.h</a></li>
<li>mapgd_0.4/math_functions/models.cc</li>
</ul>
</div><!-- contents -->
<!-- start footer part -->
<hr class="footer"/><address class="footer"><small>
Generated on Mon Oct 1 2018 13:55:21 for mapgd by &#160;<a href="http://www.doxygen.org/index.html">
<img class="footer" src="doxygen.png" alt="doxygen"/>
</a> 1.8.6
</small></address>
</body>
</html>
